src/main/java/com/example/demo/service/FoodService.java

```diff
@@ -0,0 +1,159 @@
+        existingFood.setCarbohydrates(updatedFood.getCarbohydrates());
+        existingFood.setCalories(updatedFood.getCalories());
+        return existingFood;
+    }
+
+    public void restoreDeletedFoods(Food deletedFood){
+        this.restoredFoods = new ArrayList<>();
+        for (FoodDto food : deletedFoods){
+            FoodDto restoredFood = createFood(food);
+            restoredFoods.add(restoredFood);
+        }
+    }
+
+    public void deleteFood(int foodId) {
+        //Want to save food to restore then, just if we delete it by mistake
+        Food foodToDelete = recipeRepository.findByFoodId(foodId);
+        if(foodToDelete != null) {
+            recipeRepository.deleteById(foodId);
+            foodRepository.deleteById(foodId);
+            deletedFoods.add(foodToFoodDto(foodToDelete));
+        }
+    }
+
+    public List<FoodDto> getAllFoods() {
+        List<Food> foodList = foodRepository.findAll();
+        List<FoodDto> foodDtos = new ArrayList<>();
+        for (Food food : foodList){
+            foodDtos.add(foodToFoodDto(food));
+        }
+        return foodDtos;
+    }
+
+    public List<FoodDto> searchFoodsByName(String name) {
+        List<Food> foodList = foodRepository.findByFoodName(name);
+        List<FoodDto> foodDtos = new ArrayList<>();
+        for (Food food : foodList){
+            foodDtos.add(foodToFoodDto(food));
+        }
+        return foodDtos;
+    }
+
+    public List<FoodDto> searchFoodsByProtein(double minProtein) {
+        List<Food> foodList = foodRepository.findByProteinsGreaterThanEqual(minProtein);
+        List<FoodDto> foodDtos = new ArrayList<>();
+        for (Food food : foodList){
+            foodDtos.add(foodToFoodDto(food));
+        }
+        return foodDtos;
+    }
+
+    public List<FoodDto> searchFoodsByFat(double maxFat) {
+        List<Food> foodList = foodRepository.findByFatLessThanEqual(maxFat);
+        List<FoodDto> foodDtos = new ArrayList<>();
+        for (Food food : foodList){
+            foodDtos.add(foodToFoodDto(food));
+        }
+        return foodDtos;
+    }
+
+    public List<FoodDto> searchFoodsByCarbohydrates(double minCarbohydrates){
+        List<Food> foodList = foodRepository.findByCarbohydratesGreaterThanEqual(minCarbohydrates);
+        List<FoodDto> foodDtos = new ArrayList<>();
+        for (Food food : foodList){
+            foodDtos.add(foodToFoodDto(food));
+        }
+        return foodDtos;
+    }
+
+    public List<FoodDto> searchFoodsByCalories(double minCalories) {
+        List<Food> foodList = foodRepository.findByCaloriesGreaterThanEqual(minCalories);
+        List<FoodDto> foodDtos = new ArrayList<>();
+        for (Food food : foodList){
+            foodDtos.add(foodToFoodDto(food));
+        }
+        return foodDtos;
+    }
+
+
+    public List<FoodDto> getFoodByIngredientId(int ingredientId) {
+        Ingredient ingredient = ingredientRepository.findById(ingredientId).get();

```

Comment:
If this returns Optional. then it won't be null, you will get exception here. you should use `ingredient.isPresent()`
https://www.baeldung.com/java-optional
check other usages as well